return {
	pistol = {
		{
			name = "Speed Increase",
			description = "Increases movement speed by 5% per level, up to 25%.",
			costs = { 1, 3, 5, 7, 9 },
			skillPointCost = 1,
		},
		{
			name = "Dodging",
			description = "Grants a 5% dodge chance per level, up to 50%.",
			costs = { 2, 5, 8, 12, 15, 18, 20, 22, 25 },
			skillPointCost = 2,
		},
		{
			name = "Speed Fire",
			description = "Unlocks a 60% increased firerate on all pistols.",
			costs = { 20 },
			skillPointCost = 2,
		},
		{
			name = "Stunt Bullets",
			description = "Grants a 40% chance to stun an enemy for 2 seconds.",
			costs = { 35 },
			skillPointCost = 3,
		},
	},
	sniper = {
		{
			name = "Hidden in Plain Sight",
			description = "Increases damage by 5% per level when standing still.",
			costs = { 2, 5, 8, 12, 15 },
			skillPointCost = 1,
		},
		{
			name = "Pierce Count",
			description = "Increases the number of enemies bullets can pierce by 1 per level.",
			costs = { 3, 5, 8, 10, 12 },
			skillPointCost = 1,
		},
		{
			name = "Movement Speed",
			description = "Increases your movement speed by 10%.",
			costs = { 15 },
			skillPointCost = 2,
		},
		{
			name = "Shockwave Impact",
			description = "Landing a headshot creates a shockwave that damages nearby zombies.",
			costs = { 30 },
			skillPointCost = 3,
		},
	},
	shotgun = {
		{
			name = "Pierce",
			description = "Increase pierce.",
			costs = { 2, 4, 6, 8, 10 },
			skillPointCost = 2,
		},
		{
			name = "Fight or Flight",
			description = "+10% faster fire rate per level.",
			costs = { 10, 15, 25 },
			skillPointCost = 1,
		},
		{
			name = "Boomstick",
			description = "Grants an extra pellet per shot.",
			costs = { 5, 10, 15, 20, 25 },
			skillPointCost = 2,
		},
		{
			name = "Immovable Object",
			description = "Killing an enemy grants a +10% damage increase per level for 2 seconds.",
			costs = { 10, 20, 30, 40 },
			skillPointCost = 3,
		},
	},
	rifle = {
		{
			name = "Damage Increase",
			description = "Increases damage by 2% per level, up to 40%.",
			costs = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13, 15, 15, 15, 15, 15, 15, 15, 15 },
			skillPointCost = 1,
		},
		{
			name = "Pierce",
			description = "Increases the number of enemies bullets can pierce by 1 per level, up to 5.",
			costs = { 2, 4, 6, 8, 10 },
			skillPointCost = 2,
		},
		{
			name = "Combat Conditioning",
			description = "After a kill, recover 5 health for 5 seconds. Stackable.",
			costs = { 15 },
			skillPointCost = 1,
		},
		{
			name = "Master Determination",
			description = "Grants a 30% damage increase when health is below 30%.",
			costs = { 20 },
			skillPointCost = 3,
		},
	},
	smg = {
		{
			name = "Rapid Fire",
			description = "Increases fire rate by 10% per level, up to 100%.",
			costs = { 10, 12, 13, 14, 15, 16, 17, 18, 19, 20 },
			skillPointCost = 2,
		},
		{
			name = "Suppressing Fire",
			description = "Slows zombies by 10% per level for 3 seconds.",
			costs = { 10, 20, 30, 40, 50 },
			skillPointCost = 3,
		},
		{
			name = "Piercing Shots",
			description = "Allows shots to pierce through one additional zombie per level.",
			costs = { 15, 25, 35, 45, 55 },
			skillPointCost = 1,
		},
		{
			name = "Adrenaline Surge",
			description = "Grants a 7% movement speed boost per level after getting a kill.",
			costs = { 20, 40, 55, 70, 85 },
			skillPointCost = 1,
		},
	},
}
